### PR TITLE
Reorgnize sub log

### PR DESCRIPTION
**Pull request by wangdayong228**
This change is [Reviewable](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/100)

### CODE REVIEW

**Review by Pana — CHANGES_REQUESTED** [review]
Please update the changelog, and better add a document about how to use subscribe?

> Reviewed 8 of 8 files at r1.
> *[Reviewable](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/100#-:-MfrkAhE1DW50ctJWBgv:bvmmbml)* status: all files reviewed, 1 unresolved discussion (waiting on @boqiu, @wangdayong228, and @wanliqun)

___
*[types/log.go, line 243 at r1](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/100#-Mfrj_mv2DP5Q8oGE38V:-Mfrj_mv2DP5Q8oGE38W:bmiixa1) ([raw file](https://github.com/conflux-chain/go-conflux-sdk/blob/f7079a96822ce34a40cc11047cd71c5bd370dfe0/types/log.go#L243)):*
> ```Go
> }
> 
> type rlpEncodableSubscriptionLog struct {
> ```

Why define a rlpEncodable type, is it possible to make hexutil.Big elpEncodable ?

**Review by wangdayong228 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/100)* status: 8 of 10 files reviewed, 1 unresolved discussion (waiting on @boqiu, @Pana, and @wanliqun)

___
*[types/log.go, line 243 at r1](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/100#-Mfrj_mv2DP5Q8oGE38V:-Mg54it72ihXaw7zEAhs:b-896fix) ([raw file](https://github.com/conflux-chain/go-conflux-sdk/blob/f7079a96822ce34a40cc11047cd71c5bd370dfe0/types/log.go#L243)):*
<details><summary><i>Previously, Pana (W) wrote…</i></summary><blockquote>

Why define a rlpEncodable type, is it possible to make hexutil.Big elpEncodable ?
</blockquote></details>

Done.

**Review by wangdayong228 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/100)* status: 8 of 10 files reviewed, 1 unresolved discussion (waiting on @boqiu, @Pana, and @wanliqun)

___
*[types/log.go, line 243 at r1](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/100#-Mfrj_mv2DP5Q8oGE38V:-Mg56jQE3w4o2PKfiy3e:b4mf6co) ([raw file](https://github.com/conflux-chain/go-conflux-sdk/blob/f7079a96822ce34a40cc11047cd71c5bd370dfe0/types/log.go#L243)):*
<details><summary><i>Previously, wangdayong228 wrote…</i></summary><blockquote>

Done.
</blockquote></details>

It may use by infura

File: types/log.go [hunk 1]
```diff
@@ -234,3 +234,69 @@ func resolveToHashes(val interface{}) ([]Hash, error) {
...
+}
+
+type rlpEncodableSubscriptionLog struct {
+	Log        *rlpEncodableLog        `rlp:"nil"`
+	ChainReorg *rlpEncodableChainReorg `rlp:"nil"`
+}
+
+func (s SubscriptionLog) IsRevertLog() bool {
+	return s.ChainReorg != nil
+	// return !reflect.DeepEqual(s.ChainReorg, ChainReorg{})
```
Comment (wanliqun) [inline types/log.go:250]: remove comments?

**Review by Pana — APPROVED** [review]
> Reviewed 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/100)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @boqiu and @wanliqun)

**Review by Pana — DISMISSED** [review]
> *[Reviewable](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/100)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @boqiu and @wanliqun)

File: types/log.go [hunk 2]
```diff
@@ -234,3 +234,69 @@ func resolveToHashes(val interface{}) ([]Hash, error) {
...
+}
+
+func (s SubscriptionLog) MarshalJSON() ([]byte, error) {
+	if s.IsRevertLog() {
+		return json.Marshal(s.ChainReorg)
+	}
+	return json.Marshal(s.Log)
+}
```
Comment (wanliqun) [inline types/log.go:258]: I'm wondering how the SubscribeLogs as following decodes the encoded json string? 

// SubscribeLogs subscribes all logs matching a certain filter, in order.
func (client *Client) SubscribeLogs(channel chan types.SubscriptionLog, filter types.LogFilter) (*rpc.ClientSubscription, error) {
	return client.rpcRequester.Subscribe(context.Background(), "cfx", channel, "logs", filter)
}

**Review by wangdayong228 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/100)* status: 6 of 10 files reviewed, 2 unresolved discussions (waiting on @boqiu, @Pana, and @wanliqun)

___
*[types/log.go, line 250 at r3](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/100#-MfrS-NZ_GQr2T-UZh65-r3-250:-Mg8vjqR4BtdwzefCaY1:b-896fix) ([raw file](https://github.com/conflux-chain/go-conflux-sdk/blob/9ea7fec615a8e812cb8c4c999eb98a0ebfa4398f/types/log.go#L250)):*
<details><summary><i>Previously, wanliqun (wanliqun) wrote…</i></summary><blockquote>

remove comments?
</blockquote></details>

Done.
___
*[types/log.go, line 258 at r3](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/100#-MfrS-NZ_GQr2T-UZh65-r3-258:-Mg8vFIaBa8y1Uo8-xJ1:bg9sp5g) ([raw file](https://github.com/conflux-chain/go-conflux-sdk/blob/9ea7fec615a8e812cb8c4c999eb98a0ebfa4398f/types/log.go#L258)):*
<details><summary><i>Previously, wanliqun (wanliqun) wrote…</i></summary><blockquote>

I'm wondering how the SubscribeLogs as following decodes the encoded json string? 

// SubscribeLogs subscribes all logs matching a certain filter, in order.
func (client *Client) SubscribeLogs(channel chan types.SubscriptionLog, filter types.LogFilter) (*rpc.ClientSubscription, error) {
	return client.rpcRequester.Subscribe(context.Background(), "cfx", channel, "logs", filter)
}
</blockquote></details>

json.Unmarshal is ok, please see log_test
```
sl = SubscriptionLog{}
err = json.Unmarshal([]byte(slogJson), &sl)
```

**Review by wanliqun — APPROVED** [review]

**Pull request merged**